crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -576,30 +579,108 @@ impl CheckpointBuilder {
                 }
             }
         }
-        let chunks = all_effects.chunks(self.max_transactions_per_checkpoint);
-        let chunks = chunks.into_iter().map(|ch| ch.to_vec());
-        let mut chunks: Vec<_> = chunks.collect();
-        if chunks.is_empty() {
-            // We intentionally create an empty checkpoint here if there is no content provided
+
+        let all_digests: Vec<_> = all_effects
+            .iter()
+            .map(|effect| effect.transaction_digest)
+            .collect();
+        let mut all_effects_and_transaction_sizes = Vec::with_capacity(all_effects.len());
+        for effects in all_effects {
+            let (transaction, transaction_size) = self
+                .state
+                .database
+                .get_transaction_and_serialized_size(&effects.transaction_digest)?
+                .unwrap_or_else(|| panic!("Could not find executed transaction {effects:?}"));
+            // ConsensusCommitPrologue is guaranteed to be processed before we reach here
+            if !matches!(
+                transaction.inner().transaction_data().kind,
+                TransactionKind::Single(SingleTransactionKind::ConsensusCommitPrologue(..))
+            ) {
+                // todo - use NotifyRead::register_all might be faster
+                self.epoch_store
+                    .consensus_message_processed_notify(SequencedConsensusTransactionKey::External(
+                        ConsensusTransactionKey::Certificate(effects.transaction_digest),
+                    ))
+                    .await?;
+            }
+            all_effects_and_transaction_sizes.push((effects, transaction_size));
+        }
+
+        debug!(
+            "Waiting for checkpoint user signatures for certificates {:?} to appear in consensus",
+            all_effects_and_transaction_sizes
+                .iter()
+                .map(|(effects, _)| effects)
+                .collect::<Vec<_>>()
+        );
+        let signatures = {
+            let _guard = monitored_scope("CheckpointBuilder::wait_user_signatures");
+            self.epoch_store
+                .user_signatures_for_checkpoint(&all_digests)?
+        };
+        debug!(
+            "Received {} checkpoint user signatures from consensus",
+            signatures.len()
+        );
+
+        struct CheckpointTransaction {
+            effects: TransactionEffects,
+            signatures: Vec<GenericSignature>,
+        }
+        let mut chunks: Vec<Vec<CheckpointTransaction>> = Vec::new();
+        let mut chunk: Vec<CheckpointTransaction> = Vec::new();
+        let mut chunk_size: usize = 0;
+        for ((effects, transaction_size), signatures) in all_effects_and_transaction_sizes
```

Comment:
Is it possible to convert this `for` into function and add monitored scope to measure how much time this part takes? (the monitored scope part I am most interested in)

Reply:
done